### ISSUE TITLE
Custom dataset fields causes IndexError when arbitrary other "extra" fields are used.

### ISSUE THREAD

**Issue by ntoll** [ckan/ckan#2150]
Hi,

I followed the instructions here:

http://docs.ckan.org/en/latest/extensions/adding-custom-fields.html#customizing-dataset-and-resource-metadata-fields-using-idatasetform

Unfortunately, when adding new datasets to the catalogue that have further "extra" fields I get the following error:

```
File '/home/vagrant/ckan/lib/default/src/ckan/ckan/lib/navl/dictization_functions.py', line 399 in unflatten
  current_pos = current_pos[key]
IndexError: list index out of range
```
To cut a (very) long story short, the unflatten function was getting data that included entries such as the following:

```
 ('extras',): [{'key': 'coverage_end_date', 'value': u'2002-09-01'},
               {'key': 'coverage_start_date', 'value': u'2001-09-01'},
               {'key': 'homepage', 'value': u'http://homepage.com/'},
               {'key': 'origin', 'value': u'Origin'}],
 ('extras', 0, 'key'): u'qwe',
 ('extras', 0, 'value'): u'qwe',
 ('extras', 1, 'key'): u'wer',
 ('extras', 1, 'value'): u'wer',
 ('extras', 2, 'key'): u'etr',
 ('extras', 2, 'value'): u'etr',
 ('extras', 3, 'key'): u'rty',
 ('extras', 3, 'value'): u'rty',
 ('extras', 4, 'key'): u'tyu',
 ('extras', 4, 'value'): u'tyu',
 ('extras', 5, 'key'): u'yui',
 ('extras', 5, 'value'): u'yui',
 ('extras', 6, 'key'): u'uio',
 ('extras', 6, 'value'): u'uio',
```
It turns out that the existence of the `('extras', )` key/value pair was interfering with the unflattening process of the `('extras', n, foo)` items. How did such an item come about? It turns out that `ckan.logic.converters.convert_to_extras` is the cause and doesn't work in what I would say is a safe and unsurprising manner (it generates the list of `dict` objects rather than inserts the appropriate `('extras', n, foo)` items into the data object).

Interestingly enough, the related `convert_from_extras` function *does* work in what I would call the expected manner (handling `('extras', n, foo)` items).

I'll submit a patch once I get the ticket number for this issue.

**Comment by ntoll** [comment]
Hi David,

Thanks for looking at this. I'm just back from a holiday so catching up on emails and wanted to acknowledge that I'd seen this. I'll be able to take a look at this next week once I've ploughed through backlog of other emails that need actioning.

I think the first thing I should do is make sure I write some repeatable steps to recreate the bug. Secondly, and I think I may have already mentioned this, I could be missing something due to my unfamiliarity with the internal machinations of CKAN. I've tried to avoid the "internals" and worked via the plugin system so the project will have a relatively clear/safe upgrade path. Nevertheless, I guess repeatable recreation of the bug will shine a light on either of these two possibilities.

:-)

Ciao

N.
